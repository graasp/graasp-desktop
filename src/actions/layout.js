import { MAX_APP_HEIGHT, MIN_APP_HEIGHT } from '../config/layout';
import { RADIX } from '../config/constants';
import { OPEN_TOOLS, CLOSE_TOOLS, SET_TOOLS_WIDTH } from '../types';

const openTools = () => dispatch =>
  dispatch({
    type: OPEN_TOOLS,
  });

const closeTools = () => dispatch =>
  dispatch({
    type: CLOSE_TOOLS,
  });

const setToolsWidth = ({ width }) => dispatch =>
  dispatch({
    type: SET_TOOLS_WIDTH,
    payload: width,
  });

const setHeight = (id, height) => {
  const heightToSave = Math.max(
    Math.min(height, MAX_APP_HEIGHT),
    MIN_APP_HEIGHT
  );
  localStorage.setItem(id, String(heightToSave));
};

const getHeight = id => {
  const height = localStorage.getItem(id);
  if (height) {
    return parseInt(height, RADIX);
  }
  return null;
};

export { setHeight, getHeight, openTools, closeTools, setToolsWidth };
