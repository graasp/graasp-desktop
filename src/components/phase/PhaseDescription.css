.PhaseDescriptionText p {
  font-size: x-large;
}
