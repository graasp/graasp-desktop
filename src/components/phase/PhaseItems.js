import React from 'react';
import PropTypes from 'prop-types';
import PhaseItem from './PhaseItem';
import PhaseItemDescription from './PhaseItemDescription';

const PhaseItems = ({ items, phaseId, spaceId }) => {
  if (!items) {
    return null;
  }
  return items.map(item => {
    const { description } = item;
    return (
      <>
        <PhaseItemDescription description={description} />
        <PhaseItem
          key={item.id}
          phaseId={phaseId}
          spaceId={spaceId}
          item={item}
        />
      </>
    );
  });
};

PhaseItems.propTypes = {
  items: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string,
      category: PropTypes.string,
      asset: PropTypes.string,
      url: PropTypes.string,
      name: PropTypes.string,
    })
  ),
  phaseId: PropTypes.string.isRequired,
  spaceId: PropTypes.string.isRequired,
};

export default PhaseItems;
