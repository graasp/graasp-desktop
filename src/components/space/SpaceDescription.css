.SpaceDescription {
  height: calc(100% - 64px);
  display: flex;
  justify-content: center;
  align-items: center;
}

.SpaceDescriptionDiv {
  justify-content: center;
  align-items: center;
}

.SpaceDescriptionText p {
  font-size: xx-large;
}
