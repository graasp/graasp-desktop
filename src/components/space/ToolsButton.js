import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Fab from '@material-ui/core/Fab';
import { connect } from 'react-redux';
import { withTranslation } from 'react-i18next';
import { withStyles } from '@material-ui/core/styles';
import BuildIcon from '@material-ui/icons/Build';
import CloseIcon from '@material-ui/icons/Close';
import { closeTools, openTools } from '../../actions';

class ToolsButton extends Component {
  static propTypes = {
    classes: PropTypes.shape({
      root: PropTypes.string,
      main: PropTypes.string,
      fab: PropTypes.string,
    }).isRequired,
    t: PropTypes.func.isRequired,
    dispatchOpenTools: PropTypes.func.isRequired,
    dispatchCloseTools: PropTypes.func.isRequired,
    open: PropTypes.bool.isRequired,
  };

  static styles = theme => ({
    fab: {
      margin: theme.spacing(),
      position: 'fixed',
      bottom: theme.spacing(2),
      right: theme.spacing(2),
    },
  });

  render() {
    const {
      classes,
      t,
      dispatchOpenTools,
      open,
      dispatchCloseTools,
    } = this.props;

    return (
      <Fab
        color="primary"
        aria-label={t('Settings')}
        className={classes.fab}
        onClick={open ? dispatchCloseTools : dispatchOpenTools}
      >
        {open ? <CloseIcon /> : <BuildIcon />}
      </Fab>
    );
  }
}

const mapStateToProps = ({ layout }) => ({
  open: layout.getIn(['tools', 'open']),
});

const mapDispatchToProps = {
  dispatchOpenTools: openTools,
  dispatchCloseTools: closeTools,
};

const ConnectedComponent = connect(
  mapStateToProps,
  mapDispatchToProps
)(ToolsButton);

const StyledComponent = withStyles(ToolsButton.styles)(ConnectedComponent);

export default withTranslation()(StyledComponent);
