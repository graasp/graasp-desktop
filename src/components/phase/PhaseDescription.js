import React from 'react';
import PropTypes from 'prop-types';
import Text from '../common/Text';
import './PhaseDescription.css';

const style = {
  marginBottom: '2rem',
};

const PhaseDescription = ({ description }) => {
  if (description && description !== '') {
    return (
      <Text
        content={description}
        style={style}
        className="PhaseDescriptionText"
      />
    );
  }
  return null;
};

PhaseDescription.propTypes = {
  description: PropTypes.string,
};

PhaseDescription.defaultProps = {
  description: '',
};

export default PhaseDescription;
