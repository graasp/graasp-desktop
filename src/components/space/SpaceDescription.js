import React from 'react';
import PropTypes from 'prop-types';
import Button from '@material-ui/core/Button';
import { withStyles } from '@material-ui/core';
import { useTranslation } from 'react-i18next';
import Styles from '../../Styles';
import './SpaceDescription.css';
import Banner from '../common/Banner';
import Text from '../common/Text';

const style = {
  fontSize: 'large',
};

const renderPreviewWarning = t => {
  return (
    <Banner
      type="warning"
      text={t(
        'You are previewing this space. Any input or changes will not be saved.'
      )}
    />
  );
};

const SpaceDescription = ({ description, classes, start, saved }) => {
  const { t } = useTranslation();

  return (
    <div className="SpaceDescription">
      <div>
        {saved ? null : renderPreviewWarning(t)}
        <div className={classes.spaceDescription}>
          <Text
            content={description}
            style={style}
            className="SpaceDescriptionText"
          />
        </div>
        <Button
          variant="contained"
          className={classes.button}
          onClick={start}
          size="large"
          color="primary"
          style={{ display: 'block', margin: '0 auto' }}
        >
          {saved ? t('Start') : t('Preview')}
        </Button>
      </div>
    </div>
  );
};

SpaceDescription.propTypes = {
  description: PropTypes.string.isRequired,
  classes: PropTypes.shape({ appBar: PropTypes.string.isRequired }).isRequired,
  start: PropTypes.func.isRequired,
  saved: PropTypes.bool,
};

SpaceDescription.defaultProps = {
  saved: false,
};

export default withStyles(Styles, { withTheme: true })(SpaceDescription);
