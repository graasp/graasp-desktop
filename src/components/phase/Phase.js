import React, { Component } from 'react';
import _ from 'lodash';
import PropTypes from 'prop-types';
import ImmutablePropTypes from 'react-immutable-proptypes';
import { connect } from 'react-redux';
import SplitPane from 'react-split-pane';
import { clearPhase, selectPhase, setToolsWidth } from '../../actions';
import SpaceDescription from '../space/SpaceDescription';
import PhaseDescription from './PhaseDescription';
import PhaseItems from './PhaseItems';
import ToolsButton from '../space/ToolsButton';
import './Phase.css';
import { MAX_TOOLS_WIDTH, MIN_TOOLS_WIDTH } from '../../config/layout';

const styles = {
  containerStyle: {
    position: 'static',
    height: 'calc(100% - 64px)',
  },
  mainContentStyle: {
    padding: '2rem',
    height: '100%',
    overflowY: 'scroll',
    paddingBottom: '100px',
  },
  toolContentStyle: {
    padding: '2rem',
    height: '100%',
    overflowY: 'scroll',
    backgroundColor: 'white',
    paddingBottom: '100px',
  },
};

const { containerStyle, mainContentStyle, toolContentStyle } = styles;

export class Phase extends Component {
  static propTypes = {
    space: ImmutablePropTypes.contains({
      id: PropTypes.string.isRequired,
      description: PropTypes.string,
    }).isRequired,
    phase: ImmutablePropTypes.contains({
      id: PropTypes.string,
    }).isRequired,
    dispatchSelectPhase: PropTypes.func.isRequired,
    dispatchClearPhase: PropTypes.func.isRequired,
    dispatchSetToolsWidth: PropTypes.func.isRequired,
    start: PropTypes.func.isRequired,
    toolsVisible: PropTypes.bool.isRequired,
    toolsWidth: PropTypes.number.isRequired,
  };

  componentWillUnmount() {
    const { dispatchClearPhase } = this.props;
    dispatchClearPhase();
  }

  handleResizeTools = width => {
    const { dispatchSetToolsWidth } = this.props;
    dispatchSetToolsWidth({ width });
  };

  renderContent() {
    const { phase, space } = this.props;
    const phaseDescription = phase.get('description');

    const phaseId = phase.get('id');
    const spaceId = space.get('id');
    const items = phase.get('items');
    return (
      <div style={mainContentStyle}>
        <PhaseDescription description={phaseDescription} />
        <PhaseItems items={items} spaceId={spaceId} phaseId={phaseId} />
      </div>
    );
  }

  renderWithTools() {
    const { toolsWidth, space } = this.props;

    // tools live at the top level, so the spaceId and the phaseId are the same
    // this is relevant for apps because we use these ids to locate where they
    // are saving their resources and settings
    const spaceId = space.get('id');
    const phaseId = spaceId;
    const items = space.get('items');
    return (
      <>
        <SplitPane
          split="vertical"
          maxSize={MAX_TOOLS_WIDTH}
          minSize={MIN_TOOLS_WIDTH}
          defaultSize={toolsWidth}
          primary="second"
          style={containerStyle}
          onDragFinished={this.handleResizeTools}
        >
          {this.renderContent()}
          <div style={toolContentStyle}>
            <PhaseItems items={items} spaceId={spaceId} phaseId={phaseId} />
          </div>
        </SplitPane>
      </>
    );
  }

  render() {
    const {
      phase,
      space,
      dispatchSelectPhase,
      start,
      toolsVisible,
    } = this.props;
    const phases = space.get('phases');
    const saved = space.get('saved');
    const description = space.get('description');

    // tools are items in the top level of the space
    const tools = space.get('items');

    // if there is no phase, we show the description
    // as the user has not started to use the space
    if (!phase || phase.isEmpty()) {
      return (
        <SpaceDescription
          phases={phases}
          description={description}
          selectPhase={dispatchSelectPhase}
          start={start}
          saved={saved}
        />
      );
    }

    // case where there is no tools
    if (_.isEmpty(tools)) {
      return <div style={containerStyle}>{this.renderContent()}</div>;
    }

    // case where there are tools and they are visible
    if (toolsVisible) {
      return (
        <>
          {this.renderWithTools()}
          <ToolsButton />
        </>
      );
    }

    // case where there are tools, but they are not visible
    return (
      <div style={containerStyle}>
        {this.renderContent()}
        <ToolsButton />
      </div>
    );
  }
}

const mapStateToProps = ({ Space, layout }) => ({
  space: Space.get('current').get('content'),
  toolsVisible: layout.getIn(['tools', 'open']),
  toolsWidth: layout.getIn(['tools', 'width']),
});

const mapDispatchToProps = {
  dispatchSelectPhase: selectPhase,
  dispatchClearPhase: clearPhase,
  dispatchSetToolsWidth: setToolsWidth,
};

export default connect(mapStateToProps, mapDispatchToProps)(Phase);
