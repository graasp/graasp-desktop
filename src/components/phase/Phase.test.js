import React from 'react';
import { shallow } from 'enzyme';
import { Map } from 'immutable';
import SplitPane from 'react-split-pane';
import { Phase } from './Phase';
import SpaceDescription from '../space/SpaceDescription';
import PhaseDescription from './PhaseDescription';
import PhaseItems from './PhaseItems';
import ToolsButton from '../space/ToolsButton';
import { MAX_TOOLS_WIDTH, MIN_TOOLS_WIDTH } from '../../config/layout';

const samplePhase = Map({
  id: 'phaseId',
  description: 'phaseDescription',
  items: [{ id: 'phaseItemId' }],
});

const sampleTools = [{ id: 'spaceItemsId' }];

const createPhaseProps = (phase, items, toolsVisible) => {
  return {
    space: Map({
      id: 'spaceId',
      description: 'spaceDescription',
      items,
    }),
    phase,
    dispatchSelectPhase: jest.fn(),
    dispatchClearPhase: jest.fn(),
    dispatchSetToolsWidth: jest.fn(),
    start: jest.fn(),
    toolsVisible,
    toolsWidth: 3,
  };
};

describe('<Phase />', () => {
  describe('without tools', () => {
    let wrapper;
    let props;

    beforeAll(() => {
      props = createPhaseProps(samplePhase, undefined, false);
      // eslint-disable-next-line react/jsx-props-no-spreading
      wrapper = shallow(<Phase {...props} />);
    });

    it('renders correctly', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('renders one <PhaseDescription /> component with phase description', () => {
      const phaseDescription = wrapper.find(PhaseDescription);
      expect(phaseDescription).toHaveLength(1);
      expect(phaseDescription.prop('description')).toEqual(
        props.phase.get('description')
      );
    });

    it('renders one <PhaseItems /> component with phase description', () => {
      const phaseItems = wrapper.find(PhaseItems);
      expect(phaseItems).toHaveLength(1);
      expect(phaseItems.prop('items')).toEqual(props.phase.get('items'));
      expect(phaseItems.prop('spaceId')).toEqual(props.space.get('id'));
      expect(phaseItems.prop('phaseId')).toEqual(props.phase.get('id'));
    });
  });

  describe('with tools', () => {
    describe('<Phase /> with toolsVisible = false', () => {
      let wrapper;
      let props;

      beforeAll(() => {
        props = createPhaseProps(samplePhase, sampleTools, false);
        // eslint-disable-next-line react/jsx-props-no-spreading
        wrapper = shallow(<Phase {...props} />);
      });

      it('renders correctly', () => {
        expect(wrapper).toMatchSnapshot();
      });

      it('renders one <PhaseDescription /> component with phase description', () => {
        const phaseDescription = wrapper.find(PhaseDescription);
        expect(phaseDescription).toHaveLength(1);
        expect(phaseDescription.prop('description')).toEqual(
          props.phase.get('description')
        );
      });

      it('renders one <PhaseItems /> component with phase description', () => {
        const phaseItems = wrapper.find(PhaseItems);
        expect(phaseItems).toHaveLength(1);
        expect(phaseItems.prop('items')).toEqual(props.phase.get('items'));
        expect(phaseItems.prop('spaceId')).toEqual(props.space.get('id'));
        expect(phaseItems.prop('phaseId')).toEqual(props.phase.get('id'));
      });
    });

    describe('<Phase /> with toolsVisible = true', () => {
      let wrapper;
      let props;

      beforeAll(() => {
        props = createPhaseProps(samplePhase, sampleTools, true);
        // eslint-disable-next-line react/jsx-props-no-spreading
        wrapper = shallow(<Phase {...props} />);
      });

      it('renders correctly', () => {
        expect(wrapper).toMatchSnapshot();
      });

      it('renders one <PhaseDescription /> component with phase description', () => {
        const phaseDescription = wrapper.find(PhaseDescription);
        expect(phaseDescription).toHaveLength(1);
        expect(phaseDescription.prop('description')).toEqual(
          props.phase.get('description')
        );
      });

      it('renders one <SplitPane /> component with correct properties', () => {
        const splitPane = wrapper.find(SplitPane);
        // two PhaseItems: phase and space items
        expect(splitPane).toHaveLength(1);
        expect(splitPane.prop('maxSize')).toEqual(MAX_TOOLS_WIDTH);
        expect(splitPane.prop('minSize')).toEqual(MIN_TOOLS_WIDTH);
        expect(splitPane.prop('defaultSize')).toEqual(props.toolsWidth);
        expect(splitPane.find(PhaseItems)).toHaveLength(2);
      });

      it('renders one <PhaseItems /> component with phase content', () => {
        const phaseItems = wrapper
          .find(PhaseItems)
          .find({ items: props.phase.get('items') });
        expect(phaseItems).toHaveLength(1);
        expect(phaseItems.prop('spaceId')).toEqual(props.space.get('id'));
        expect(phaseItems.prop('phaseId')).toEqual(props.phase.get('id'));
      });

      it('renders one <PhaseItems /> component with space content (tools)', () => {
        const phaseItems = wrapper
          .find(PhaseItems)
          .find({ items: props.space.get('items') });
        expect(phaseItems).toHaveLength(1);
        expect(phaseItems.prop('spaceId')).toEqual(props.space.get('id'));
        expect(phaseItems.prop('phaseId')).toEqual(props.space.get('id'));
      });

      it('renders one <ToolsButton /> component', () => {
        const toolsButton = wrapper.find(ToolsButton);
        expect(toolsButton).toHaveLength(1);
      });
    });
  });

  describe.each([undefined, Map({})])(
    '<Phase /> when phase is undefined or empty',
    phase => {
      it('renders <SpaceDescription />', () => {
        const props = createPhaseProps(phase, sampleTools, false);
        // eslint-disable-next-line react/jsx-props-no-spreading
        const wrapper = shallow(<Phase {...props} />);
        expect(wrapper.find(SpaceDescription)).toHaveLength(1);
      });
    }
  );
});
