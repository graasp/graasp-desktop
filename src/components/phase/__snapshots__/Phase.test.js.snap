// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Phase /> with tools <Phase /> with toolsVisible = false renders correctly 1`] = `ShallowWrapper {}`;

exports[`<Phase /> with tools <Phase /> with toolsVisible = true renders correctly 1`] = `ShallowWrapper {}`;

exports[`<Phase /> without tools renders correctly 1`] = `ShallowWrapper {}`;
